### PR TITLE
feat: Java field resolution

### PR DESCRIPTION
**Pull request by JonathanStarup**
Also fixes `isReducible`

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 1]
```diff
@@ -311,6 +304,22 @@ object TypeReduction {
     }
   }
 
+  private def retrieveField(clazz: Class[_], fieldName: String, loc: SourceLocation): JavaFieldResolutionResult = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:307]: Add comments

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 2]
```diff
@@ -311,6 +304,22 @@ object TypeReduction {
...
+      getField(clazz, fieldName) match {
+        case Some(field) => JavaFieldResolutionResult.Resolved(Type.Cst(TypeConstructor.JvmField(field), loc))
+        case None => JavaFieldResolutionResult.FieldNotFound
+      }
+  }
+
+  private def classFromFlixType(tpe: Type): Option[Class[_]] = tpe match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:314]: Add comments

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:314]: How is this related to Type.getFlixType?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 3]
```diff
@@ -335,6 +344,15 @@ object TypeReduction {
     }
   }
 
+  private def getField(clazz: Class[_], fieldName: String): Option[Field] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:347]: Add comments

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:347]: Can this be put next to the other retrieveField or?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 4]
```diff
@@ -335,6 +344,15 @@ object TypeReduction {
...
+    if (clazz.isArray && fieldName == "length") None // TODO this should return some of the length field
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:348]: Huh?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala [hunk 5]
```diff
@@ -335,6 +344,15 @@ object TypeReduction {
...
+    else try {
+      Some(clazz.getField(fieldName))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala:350]: What about field visibility?

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala (hunk 3 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala]: I will make a follow up pr to this file - remove unused variables, rearrange methods, etc

File: main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala (hunk 5 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala]: The method code doesn't have anything checking for that, but good point. follow-up PR?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala]: OK

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**